sql/catalyst/src/main/scala/org/apache/spark/sql/types/DecimalType.scala

```diff
@@ -89,6 +90,7 @@ case class DecimalType(precision: Int, scale: Int) extends FractionalType {
       (precision - scale) <= (dt.precision - dt.scale) && scale <= dt.scale
     case dt: IntegralType =>
       isTighterThan(DecimalType.forType(dt))
+    // For DoubleType/FloatType, the value can be NaN, PositiveInfinity or NegativeInfinity.
```

Comment:
For `isWiderThan`, yes it's not safe to cast float/double to decimal because of NaN stuff.

For `isTighterThan`, I think it's safe to cast decimal to float/double if the precision doesn't exceed?

Reply:
> For isTighterThan, I think it's safe to cast decimal to float/double if the precision doesn't exceed?

Yes, it is. I was about to push the commit to fix tests.